torch/_refs/__init__.py

```diff
@@ -3771,17 +3810,25 @@ def eye(
     # result.requires_grad_(requires_grad)
 
 
-# TODO: missing kwargs (e.g. layout)
 @out_wrapper()
 def full(
     shape: ShapeType,
     fill_value: NumberType,
     *,
-    dtype: torch.dtype,
-    device: torch.device,
-    requires_grad: bool,
+    dtype: Optional[torch.dtype] = None,
+    layout: Optional[torch.layout] = None,
+    device: Optional[torch.device] = None,
+    pin_memory: bool = False,
+    requires_grad: bool = False,
 ) -> TensorLikeType:
-    e = empty(shape, dtype=dtype, device=device, requires_grad=requires_grad)
+    e = empty(
```

Comment:
Doesn't the dtype passed to `empty()` need to be inferred from the `fill_value` type?

Reply:
pre-existing condition

Reply:
torch.full is respecting the dtype value, instead of inferring from the fill_value's dtype.

```
>>> torch.full((3,), False, dtype=torch.float)
tensor([0., 0., 0.])
```

Reply:
The question here was what to do when dtype is not passed. `full` infers dtype from fill value, whereas `empty` would create default dtype tensor when dtype is None (and thus `full` result will be wrong)

Reply:
ic... 
but why we don't we directly use prims.full? Is there any particular reason we are doing empty+fill?

Reply:
we could, but device/dtype etc args in prims.full are non-optional, so they still need to be set (btw @mruberry we probably need to assert that those args are set in prims?)

Reply:
They are required args when calling the prim:

https://github.com/pytorch/pytorch/blob/4db1588ca09b4f6328c5bd98701e9c302cb39800/torch/_prims/__init__.py#L2357

Using `full` directly sounds better than calling `empty` and then `fill` in most cases, there are some caveats with modeling new functions that we may or may not be able to model now, however. The point @ngimel and I are making is that the metadata for the `prims.full` call needs to be inferred by the reference.